### PR TITLE
Support building with CMake

### PR DESCRIPTION
**Pull request by reuk**
This pull request adds support for building with CMake. It transpired that not all of the source files in the repository were being built by the makefile build, so these extra files have been removed.

### CODE REVIEW

**Comment by NathanJPhillips** [comment]
This issue was discussed briefly at last Friday's retrospective with @jgwilson42 and has widespread support for various reasons including the fact it gives us the ability to automatically generate project files for different build systems across multiple platforms, which will simplify setting up a build environment for both new employees and external collaborators.
This PR now needs updating for the new repository format, this will take a small but finite amount of time, perhaps a general sense of whether this PR is likely to be approved once complete from @jgwilson42 or @peterschrammel would be useful before doing embarking on this work.

**Comment by jgwilson42** [comment]
We will need to prove that the output from make and cmake is functionally equivalent in all of the different builds. Any suggestions for how to do this?

**Comment by reuk** [comment]
I fixed this up so that it builds again. My suggestion would be to use the existing regression tests to compare the functionality of the different builds.

**Comment by tautschnig** [comment]
Can anyone describe in more detail as to what is perceived as problematic/painful for people with the existing set up?

File: scripts/cbmc.spec [hunk 1]
```diff
@@ -1,56 +0,0 @@
-Name:   cbmc
```
Comment (tautschnig) [inline scripts/cbmc.spec]: Why is this file being removed?

File: scripts/glucose-syrup-patch [hunk 2]
```diff
@@ -1,127 +0,0 @@
-diff -rupN glucose-syrup/core/Solver.cc glucose-syrup-patched/core/Solver.cc
```
Comment (tautschnig) [inline scripts/glucose-syrup-patch]: Why is this file being removed?

File: scripts/make-rpm [hunk 3]
```diff
@@ -1,9 +0,0 @@
-#!/bin/bash
```
Comment (tautschnig) [inline scripts/make-rpm]: Why is this file being removed?

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+#   -[ ] Specify one of many different solver libraries.
+#   -[x] Code signing.
+#   -[x] Cprover jar build.
+
+project(CBMC)
+
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD_REQUIRED true)
+
+set(CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS} -Wno-deprecated-register)
```
Comment (tautschnig) [inline src/CMakeLists.txt]: What about all the other bits that config.inc and common used to set up?

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    CMAKE_ARGS -DCMAKE_INSTALL_PREFIX:PATH=<INSTALL_DIR>
+    BUILD_BYPRODUCTS ${minisat_lib}
+)
+
+add_library(minisat2-condensed STATIC IMPORTED)
+set_target_properties(minisat2-condensed PROPERTIES
+    IMPORTED_LOCATION ${minisat_lib}
+    INTERFACE_INCLUDE_DIRECTORIES "${extern_include_directory}"
+)
+add_dependencies(minisat2-condensed minisat2-extern)
```
Comment (tautschnig) [inline src/CMakeLists.txt]: What about Glucose?

**Comment by reuk** [comment]
- No direct support for Visual Studio, CLion, Xcode, Ninja
- Cross-platform support (file extensions + locations) has to be manually maintained - CMake handles this on its own
- Builds are not out-of-source, at least by default
- Build is not one-step (minisat2-download isn't automatic) so it's easy to forget part of the build (like with the jar support via libzip, although miniz fixes this)
- No progress readouts - we can't see how long the build is expected to take
- Difficult to maintain - makefiles are difficult to debug if you don't have experience with them, whereas CMake configs look *more* like a procedural language

This isn't an exhaustive list. The most important point is support for many different build back-ends across all major platforms.

File: src/ansi-c/Makefile [hunk 6]
```diff
@@ -1,112 +0,0 @@
...
-
-file_converter$(EXEEXT): file_converter.cpp
-	$(LINKNATIVE)
-
-ifeq ($(BUILD_ENV),MinGW)
-  platform_unavail = library/java.io.c library/err.c library/threads.c
-else
-  platform_unavail = library/java.io.c library/threads.c
-endif
-library_check: library/*.c
```
Comment (tautschnig) [inline src/ansi-c/Makefile]: This target is missing

**Comment by tautschnig** [comment]
I completely admit that I may have grown up in a different age and thus probably
my opinion may be completely backwards (literally):

On Wed, Mar 08, 2017 at  3:27:09 -0800, Reuben Thomas wrote:
> - No direct support for Visual Studio, CLion, Xcode, Ninja

Ok, if people need that; though note that this will imply that all those project
files are generated and stored in the repository. Else people first need cmake.
See below.

> - Cross-platform support (file extensions + locations) has to be manually
> maintained - CMake handles this on its own

This seemed like a one-off effort, if I'm right?

> - Builds are not out-of-source, at least by default

Accepted, even though this has been worked around.

> - Build is not one-step (minisat2-download isn't automatic) so it's easy to
> forget part of the build (like with the jar support via libzip, although miniz
> fixes this)

If people wanted it to be one-step, then this would be trivial to fix by just
adding in a dependency to the "all" rule.

> - No progress readouts - we can't see how long the build is expected to take

Sure, if people are keen on that.

> - Difficult to maintain - makefiles are difficult to debug if you don't have
> experience with them, whereas CMake configs look *more* like a procedural
> language
> 

I must say that I perceive it as the complete opposite, but then I've grown up
with Makefiles.

> This isn't an exhaustive list. The most important point is support for many
> different build back-ends across all major platforms.
> 

I've been burnt too many times by the ever-so-portable build systems. There are
several pieces of software I can't build on the systems I have to work with
simply because I can't fulfil the demands for cmake. (Just checked: one system
I'm working with can offer cmake 2.6, another is Ubuntu 14.04 LTS - which has
cmake 2.8, as is true for Debian stable). Or automake/autoconf, which I've been
a heavy (and happy) user of. It all looks so beautiful when you're on some
recent Linux distro. Mac OS doesn't even ship cmake.

Please do keep in mind who may be using and building CBMC. It's great when the
development team is having a good time. But in the end it's the eventual
customers that should matter the most.

Best,
Michael

**Comment by NathanJPhillips** [comment]
My conclusions from a lively debate over lunch are that this will make a small difference in the DiffBlue office but may make the difference between people adopting or not adopting CBMC in the open source community. CMake will allow, in particular, people with a Windows dev box to contribute to the project. The pain of setting up the current system to build on Windows means that any such potential contributors at the moment will almost definitely not use or contribute to CMBC. In the professional developer community at large (though much less so in the academic community) this could be a significant number of people that are passing over our project. At least two people around at lunchtime had tried and given up building CBMC on Windows.

**Comment by forejtv** [comment]
As for Windows setup, having gone through setting up AppVeyor I share the pain, but I don't see how cmake will help. For me the biggest pain was setting up the prerequisites (flex, bison...), the rest was relatively straightforward. I suspect that cmake won't help in setting up the prerequisites, or will it?

**Comment by martin-cs** [comment]
On Wed, 2017-03-08 at 05:34 -0800, Nathan Phillips wrote:
> My conclusions from a lively debate over lunch are that this will make a small difference in the DiffBlue office but may make the difference between people adopting or not adopting CBMC in the open source community.

This is a factor but so far I have not encountered it happening in my
experience of helping people use CPROVER.

>  CMake will allow, in particular, people with a Windows dev box to contribute to the project. The pain of setting up the current system to build on Windows means that any such potential contributors at the moment will almost definitely not use or contribute to CMBC.

We have supported people developing on Windows before and the pain
points seem to be more with things like flex and bison.  For these
people tend to install Cygwin or similar (Linux VMs on windows or OS X
seem to be favoured by some people) which then means they likely have
make.

>  In the professional developer community at large (though much less so in the academic community) this could be a significant number of people that are passing over our project. At least two people around at lunchtime had tried and given up building CBMC on Windows.

Could they possibly file patches for:

https://github.com/diffblue/cbmc/blob/master/COMPILING


If you are concerned about cross platform issues and users on Windows,
the *FAR* bigger problem (I have lost count of the number of times I
have worked with people having this issue) is having to run CBMC from
the Visual Studio command prompt.  Doing something about that, would, in
my experience, make a far bigger difference to the user experience and
up-take on Windows than build system changes as most Windows users
simply download the binary version from the website.

**Comment by martin-cs** [comment]
<snip>
> > - Build is not one-step (minisat2-download isn't automatic) so it's easy to
> > forget part of the build (like with the jar support via libzip, although miniz
> > fixes this)
> 
> If people wanted it to be one-step, then this would be trivial to fix by just
> adding in a dependency to the "all" rule.

Not saying that it should, or shouldn't be a particular way but the
reason for this is that make minisat2-download was a much more recent
addition to the build system.  Historically, not everyone build with
MiniSAT (the options to build with other SAT solvers are still there and
should still work) and not everyone built with that version of MiniSAT.
We added the download option because for a while the latest version of
MiniSAT required patches for a clean build.

**Comment by tautschnig** [comment]
As I've got a bit of a blind spot for Windows: I'd like to make sure that symptoms and causes are clearly identified. What are the fatal hurdles such that "At least two people around at lunchtime had tried and given up building CBMC on Windows?" Is it the missing Visual Studio build environment? If so, probably the Visual-Studio scripts should be fixed or removed irrespective of the further progress of this pull request.

**Comment by peterschrammel** [comment]
I don't see any priority in pursuing this, except that I think that it is generally useful to investigate alternative build systems. It's clear that this can only go live when it has at least those features that we have at the moment and none of the current users is being disadvantaged. As pointed out above, if someone wants to help Windows users avoid cygwin it would be worth looking into ports of flex/bison and getting the necessary VS environment variables set automatically when running cbmc from a 'normal' command prompt.

**Comment by tautschnig** [comment]
I'm aware this isn't the full solution, but here's a script that we had used in past experiments on a Windows machine, ssh-ing into the cygwin shell:
```
# From http://stackoverflow.com/a/3272301 with minor adaption
# These lines will be installation-dependent.
export VSINSTALLDIR='C:\Program Files (x86)\Microsoft Visual Studio 10.0\'
export WindowsSdkDir='C:\Program Files\Microsoft SDKs\Windows\v7.0A\'
export FrameworkDir='C:\WINDOWS\Microsoft.NET\Framework\'
export FrameworkVersion=v4.0.30319
export Framework35Version=v3.5

# The following should be largely installation-independent.
export VCINSTALLDIR="$VSINSTALLDIR"'VC\'
export DevEnvDir="$VSINSTALLDIR"'Common7\IDE\'

export FrameworkDIR32="$FrameworkDir"
export FrameworkVersion32="$FrameworkVersion"

export INCLUDE="${VCINSTALLDIR}INCLUDE;${WindowsSdkDir}include;"
export LIB="${VCINSTALLDIR}LIB;${WindowsSdkDir}lib;"
export LIBPATH="${FrameworkDir}${FrameworkVersion};"
export LIBPATH="${LIBPATH}${FrameworkDir}${Framework35Version};"
export LIBPATH="${LIBPATH}${VCINSTALLDIR}LIB;"

c_VSINSTALLDIR=`cygpath -ua "$VSINSTALLDIR\\\\"`
c_WindowsSdkDir=`cygpath -ua "$WindowsSdkDir\\\\"`
#c_FrameworkDir=`cygpath -ua "$FrameworkDir\\\\"`

export PATH="${c_WindowsSdkDir}bin:$PATH"
export PATH="${c_WindowsSdkDir}bin/NETFX 4.0 Tools:$PATH"
export PATH="${c_VSINSTALLDIR}VC/VCPackages:$PATH"
export PATH="${c_FrameworkDir}${Framework35Version}:$PATH"
export PATH="${c_FrameworkDir}${FrameworkVersion}:$PATH"
export PATH="${c_VSINSTALLDIR}Common7/Tools:$PATH"
export PATH="${c_VSINSTALLDIR}VC/BIN:$PATH"
export PATH="${c_VSINSTALLDIR}Common7/IDE/:$PATH"
```

It may make sense to just add a few entries (various Visual Studio versions' default paths) to the PATH before trying to invoke cl.exe from the C front-end. Input from Windows users much appreciated.

**Comment by NathanJPhillips** [comment]
I had a look into using the new Visual C++ for Linux Development extension with CBMC as an alternative approach this morning but didn't get it working.

**Comment by tautschnig** [comment]
So is the missing Visual Studio set up what mainly stopped people from getting work done on Windows? It would help quite a lot to understand in detail what people are being blocked by.

**Pull request closed without merging**